cmd/specdiff/differ.go

```diff
@@ -0,0 +1,283 @@
+		}
+		if endpoint, ok := childNodeKey.Interface().(diff.Endpoint); ok {
+			changePath.fieldPath.push(handleEndpointStruct(endpoint))
+			err := searchNode(childNode, diffProto, changePath)
+			if err != nil {
+				return err
+			}
+			changePath.fieldPath.pop()
+			continue
+		}
+		return fmt.Errorf("map node key %v is not supported", childNodeKey)
+
+	}
+	return nil
+}
+
+func searchArrayAndSliceType(arrayNode reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	if arrayNode.Kind() != reflect.Slice && arrayNode.Kind() != reflect.Array {
+		panic("searchArrayAndSliceType called with invalid type")
+	}
+
+	for i := 0; i < (arrayNode.Len()); i++ {
+		childNode := arrayNode.Index(i)
+		if childNode.IsZero() {
+			continue
+		}
+		if childNode.Kind() == reflect.String {
+			changePath.fieldPath.push(childNode.String())
+			addToDiffProto(diffProto, changePath)
+			changePath.fieldPath.pop()
+			continue
+		}
+		if endpoint, ok := childNode.Interface().(diff.Endpoint); ok {
+			changePath.fieldPath.push(handleEndpointStruct(endpoint))
+			addToDiffProto(diffProto, changePath)
+			changePath.fieldPath.pop()
+			continue
+		}
+		return fmt.Errorf("array child node %v is not supported", childNode)
+	}
+	return nil
+}
+
+func searchStructType(structNode reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	if structNode.Kind() != reflect.Struct {
+		panic("searchStructType called with invalid type")
+	}
+
+	if vd, ok := structNode.Interface().(diff.ValueDiff); ok {
+		handleValueDiffStruct(vd, diffProto, changePath)
+		return nil
+	}
+	for i := 0; i < structNode.NumField(); i++ {
+		tag, ok := structNode.Type().Field(i).Tag.Lookup("json")
+		if !ok {
+			// Fields that don't have a JSON name aren't part of the diff.
+			continue
+		}
+
+		// Empty fields in the diff are redundant. Skip them.
+		childNode := structNode.Field(i)
+		if childNode.IsZero() {
+			continue
+		}
+
+		// Struct field tags have the format "fieldname,omitempty". We only want the field name.
+		fieldName := strings.Split(tag, ",")[0]
+
+		err := handleStructField(childNode, fieldName, diffProto, changePath)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func handleStructField(value reflect.Value, name string, diffProto *rpc.Diff, changePath *change) error {
+	// Empty fields in the diff are redundant. Skip them.
+	if value.IsZero() {
+		return nil
+	}
+
+	switch name {
+	case "added", "deleted", "modified":
+		changePath.changeType = name
+	default:
+		changePath.fieldPath.push(name)
+		defer changePath.fieldPath.pop()
+	}
+
+	err := searchNode(value, diffProto, changePath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleEndpointStruct(ed diff.Endpoint) string {
+	Method := getAtomicType(reflect.ValueOf(ed.Method))
+	Path := getAtomicType(reflect.ValueOf(ed.Path))
+	return fmt.Sprintf("{method.%s path.%s}", Method, Path)
+}
+
+func handleValueDiffStruct(vd diff.ValueDiff, diffProto *rpc.Diff, changePath *change) {
```

Comment:
I think the code might be clearer if this function is removed. It's only used in one place, and can be handled in one line instead:
```
diffProto.Modification[changePath.fieldPath.String()] = &rpc.Diff_ValueChange{
	From: getAtomicType(reflect.ValueOf(vd.From)),
	To:   getAtomicType(reflect.ValueOf(vd.To)),
}
```